packages/loopring_v3/contracts/amm/libamm/AmmJoinProcess.sol

```diff
@@ -83,109 +37,178 @@ library AmmJoinProcess
-            }
+        bool mintToLayer1 = (
+            join.direction == AmmData.Direction.L1_TO_L1 ||
+            join.direction == AmmData.Direction.L2_TO_L1);
 
-            ctx.ammExpectedL2Balances[i] = ctx.ammExpectedL2Balances[i].add(amount);
+        if (mintToLayer1) {
+            S.mint(join.owner, mintAmount);
+        } else {
+            _approvePoolTokenDeposit(ctx, mintAmount, join.owner);
         }
+    }
+
+    // TODO（daniel): optimize this using 1 deposit + multiple L2 transfers for some cases?
+    function _approvePoolTokenDeposit(
+        AmmData.Context  memory  ctx,
+        uint96                   amount,
+        address                  to
+        )
+        private
+    {
+        require(amount > 0, "INVALID_DEPOSIT_AMOUNT");
 
-        S.mint(address(this), poolAmountOut);
-        S.addUserBalance(address(this), join.owner, poolAmountOut);
+        // Check that the deposit in the block matches the expected deposit
+        DepositTransaction.Deposit memory deposit = ctx._block.readDeposit(ctx.txIdx++);
+        ctx.numTransactionsConsumed++;
+
+        AmmData.Token memory poolToken = ctx.tokens[ctx.size];
+
+        // Question(brecht):can we read the `to` address?
+        require(
+            deposit.owner == address(this) &&
+            deposit.accountID== ctx.accountID &&
+            deposit.tokenID == poolToken.tokenID &&
+            deposit.amount == amount,
+            "INVALID_TX_DATA"
+        );
+
+        ctx.exchange.deposit{value: 0}(
+            address(this),
+            to,
+            poolToken.addr,
+            amount,
+            new bytes(0)
+        );
     }
 
     function _calculateJoinAmounts(
-        AmmData.State    storage S,
         AmmData.Context  memory  ctx,
         AmmData.PoolJoin memory  join
         )
         private
         view
         returns(
-            bool            slippageRequirementMet,
-            uint            poolAmountOut,
+            bool            slippageOK,
+            uint96          mintAmount,
             uint96[] memory amounts
         )
     {
         // Check if we can still use this join
-        amounts = new uint96[](ctx.size);
-        uint _totalSupply = S.totalSupply;
+        amounts = new uint96[](ctx.size - 1);
 
         if (block.timestamp > join.validUntil) {
             return (false, 0, amounts);
         }
 
-        if (_totalSupply == 0) {
-            return(true, ctx.poolTokenInitialSupply, join.maxAmountsIn);
+        if (ctx.totalSupply == 0) {
+            return(true, ctx.poolTokenInitialSupply.toUint96(), join.joinAmounts);
         }
 
-        // Calculate the amount of liquidity tokens that should be minted
-        for (uint i = 0; i < ctx.size; i++) {
+        // Calculate the amount of pool tokens that should be minted
+        bool initialized = false;
+        for (uint i = 1; i < ctx.size; i++) {
```

Comment:
`1` -> `0`?

Reply:
I unresolved a couple of these, but I'm actually not sure how you want this to be stored in the arrays now. I think there's definitely some inconsistency because in some places the index can go out of bounds.

Reply:
The last element of the array is the pool token.